realm/realm-library/src/main/java/io/realm/internal/SharedGroup.java

```diff
@@ -293,6 +297,29 @@ public int hashCode() {
         }
     }
 
+    /**
+     * Waits for change committed by {@link SharedGroup} in other Thread.
+     *
+     * @return {@code true} if successfully detects change, {@code false} no change has detected otherwise.
+     */
+    public boolean waitForChange() {
+        if (isStoppedWaitingByUser) {
+            isStoppedWaitingByUser = false;
+            return false;
+        }
```

Comment:
if they are both locked on the same lock, then `stopWaitForChange` will wait forever when `waitForChange` get called first.